containers/ecr-viewer/src/app/services/auditLogService.ts

```diff
@@ -21,51 +22,86 @@ type Action =
   | "signin"
   | "signout";
 
-type AuditableFn<Params extends Record<string, unknown>, Ret> = (
+const auditTransaction = async <Params extends Record<string, unknown>, Ret>(
+  subject: Subject,
+  action: Action,
   params: Params,
-  trx: Transaction<Core>,
-) => Promise<Ret>;
+  executeFn: (trx: Transaction<Core>) => Promise<Ret>,
+): Promise<Ret> => {
+  return await getDb<Core>()
+    .transaction()
+    .execute(async (trx) => {
+      const result = await executeFn(trx);
+      const logParams =
+        typeof result === "string" ? { ...params, uuid: result } : params;
+
+      try {
+        await createAuditRecord<Params>(trx, subject, action, logParams);
+      } catch (error: unknown) {
+        if (await dbIsValid()) {
+          throw error;
+        } else {
+          console.warn({
+            message: "Audit logging failed as db is not in a valid state",
+            error,
+          });
+        }
+      }
+      return result;
+    });
+};
 
 /**
  * Wrap a function with audit logging. After the function successfully runs, an
  * audit log record will be created with the subject, action, actor (user uuid or token),
  * parameters passed to the function, and other request metadata.
+ *
+ * This is for functions that do not query the database, e.g. `getFhirData()`.
+ * @param subject Subject of the action being audited (e.g. "ecr")
+ * @param action Action being done (e.g. "view")
+ * @param fn Function to audit upon successful completion. Must be called with only one argument, which is an object with all of the parameters. If it returns a string, it is assumed to be the `uuid` of the subject and will be added to the params as such. The wrapper will inject the second argument of a Kysely transaction, which should be used as the database in any queries the function executes
+ * @returns Wrapped function
+ */
+export const auditWithoutTrx = <Params extends Record<string, unknown>, Ret>(
+  subject: Subject,
+  action: Action,
+  fn: (params: Params) => Promise<Ret>,
+) => {
+  return async (params: Params): Promise<Ret> => {
+    if (!dbDialect()) {
+      return await fn(params);
+    }
```

Comment:
Could we have added this at the beginning of the original `audit` function and gotten to the same place with a lot less change? I'm waffling on whether the added functions/explicitness is worth the clarity here